### PR TITLE
Martial Arts Variable Moves Per Weapon (and reach techniques)

### LINKED ISSUE #59585: Martial arts compatible with long weapons do not trigger techniques on reach attacks
### Describe the bug

Techniques do not trigger when reach attacks are used.

This reduces the usefulness of relevant martial arts considerably.

Especially due to the fact that a large percentage of reach weapons have 30% damage penalty when used for non-reach (adjacent) attacks.

So, for example using a halberd, naginata, or pike, you are currently stuck either with 30% damage penalty or non-working techniques.

### Steps to reproduce

1. Make a new character with "Melee Training" background and select Pencak Silat
2. Level up Melee to 5, switch to Silat in the "_" menu
3. Spawn a pike and some zombies
4. Attack with reach and long reach attacks, either with f key or Tab key.
5. Notice that no matter how much you attack, none of the techniques trigger.

### Expected behavior

I expect martial arts techniques to trigger on reach attacks.

### Screenshots

not_very_silat

### Versions and configuration

Game version: cdda-linux-tiles-x64-2022-07-22-0050
Mods: default

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bombasticSlacks**
#### Summary
Features "Martial Arts Variable Moves Per Weapon (and reach techniques)"

#### Purpose of change
_**This will be extended next release into a larger standardizing of melee**_

fixes #59585

As discussed on discord a way to do split moves would be to add techniques to the weapons themselves with different damage techniques.

This is just a proof of concept and the POKE example should be removed before this can be merged since it currently breaks stuff like the halberd that only does cut damage.

#### Describe the solution
This is the basics of the infrastructure required to do this overhaul well and easily. 

IN THIS PR:
- Unhardcode no reach techniques 
- Set all techniques to default to non reaching
- When selecting techniques filter selection based on if reach attacking 
- Weapon categories can have default techniques added to all weapons in the category
- Adds example polearm technique POKE which does only pierce damage

TODO (But out of scope for me right now)
- Weapons need a generalized *damage*/*weighting* value (maybe derived from just weight / descriptors) that techniques can scale and transform the damage type of for this to really work
- Pierce should probably stop being a flag
- Someone needs to make the techniques for all the weapons
- Someone needs to add them to the weapon_categories
- Someone needs to do a pass on weapons that don't have categories (like the Lucern Hammer)

**TLDR bunch of JSON needs to be done with more care than I'm willing to put in, and whoever does it needs to decide how they want to derive their core value this is all based on for melee weapons**

#### Testing
image

image

#### Additional context
Leaving this up as a proof of concept someone can adopt.

### CODE REVIEW

**Comment by catdach** [comment]
I'm not sure that I fully understand what this is trying to do. Is the goal just to have weapon techniques attached to weapon categories (rather than the individual weapons)? I get the feeling that something is going over my head.


> - Someone needs to do a pass on weapons that don't have categories (like the Lucern Hammer)

That's doable, with some caveats. So far weapon categories have only been added if there is a martial art that uses it. So weapons that aren't used in any martial arts generally don't have a weapon category. Basically if there isn't a use case for the category then it generally doesn't exist so that we don't bloat the number weapon categories. That being said I have no idea why the Lucern Hammer isn't in the polearm category lol

I'll go through the files in \json\items\melee to see if there are any other obvious ones that were missed, but I'm not going to make a category for baseball bats or anything like that.

**Comment by bombasticSlacks** [comment]
This was discussed on discord as a solution to splitting damage types. So a sword could hilt bash for bash damage, thrust for pierce and slash for cut for example. Something like a Lucerne hammer could have only pierce at range, etc.

Basically this is just the infrastructure, someone would need to come in and actually add all the json for it. I just was looking to verify I was right about the infrastructure being pretty trivial to add, so I did it.

**Comment by Termineitor244** [comment]
Sounds great! But this gives me the feeling that there will always be more work to be done with the weapon categories haha (I was expecting to see then finished with the MMA ones, but it seems there will be more work to do).

**Comment by Zireael07** [comment]
Likely needs a rebase after the recent MA fixes

**Comment by Terrorforge** [comment]
One thing to consider when assigning techniques to weapon groups is how this will interact with weapons that are in multiple groups. Like if a weapon like the poleaxe ends up in a hypothetical AXES and CUTTING_POLEARMS categories, it could end up with duplicate cutting techniques.

Also I want to say I think this is a de facto move towards more genericized martial arts. There's been a lot of debate re: if categories like "japanese swords" actually make sense or if you should be able to do Niten Ichi-Ryu with any two-handed sword capable of cutting. This approach is going to necessitate creating categories like "two-handed cutting sword", and if those exist, people re going to want to use them in martial arts. Not saying that's a problem, just a thought.

**Comment by bombasticSlacks** [comment]
> it could end up with duplicate cutting techniques.

Weapons will have black lists, I'll prune duplicates and this would also be associated with a move to as your character progresses in melee skill they will basically roll multiple moves and choose the "best" probably with a selectable toggle for "fight agressively" or "fight defensively" though that toggle may be martial art dependent too.

**Comment by bombasticSlacks** [comment]
Partially implemented in other PRs, otherwise too large of a rewrite for vanilla
